File: go-backend-interview-prep/10_design_patterns_in_go/09_builder_vs_functional_options/main.py
"""Builder vs functional options -- Python equivalent of the Go example.

Python's **kwargs with defaults solves the same problem natively.
"""

import time


class PrintSender:
    def send(self, to: str, message: str) -> None:
        print(f"  [SEND] {to}: {message}")


# ============================================================
# APPROACH 1: kwargs with defaults (Python-idiomatic)
# ============================================================

class NotificationService:
    """Python solves 'many optional params' with kwargs + defaults."""

    def __init__(
        self,
        sender,
        timeout: float = 5.0,
        retries: int = 1,
        verbose: bool = False,
    ):
        self.sender = sender
        self.timeout = timeout
        self.retries = retries
        self.verbose = verbose

    def notify(self, to: str, msg: str) -> None:
        if self.verbose:
            print(f"  [VERBOSE] timeout={self.timeout}s retries={self.retries}")
        print(f"  Sending to {to}: {msg}")


# ============================================================
# APPROACH 2: Builder (verbose in Python, shown for comparison)
# ============================================================

class ServiceBuilder:
    def __init__(self, sender):
        self._sender = sender
        self._timeout = 5.0
        self._retries = 1
        self._verbose = False

    def timeout(self, t: float):
        self._timeout = t
        return self

    def retries(self, n: int):
        self._retries = n
        return self

    def verbose(self):
        self._verbose = True
        return self

    def build(self) -> NotificationService:
        return NotificationService(
            self._sender,
            timeout=self._timeout,
            retries=self._retries,
            verbose=self._verbose,
        )


def main() -> None:
    print("=== Builder vs Functional Options ===")
    print()
    sender = PrintSender()

    # kwargs with defaults (natural)
    print("--- kwargs with defaults (Python-idiomatic) ---")

    svc1 = NotificationService(sender)  # all defaults
    print("  defaults: ", end="")
    svc1.notify("alice", "hello")

    svc2 = NotificationService(sender, timeout=10, retries=3, verbose=True)
    print("  custom:   ", end="")
    svc2.notify("bob", "world")
    print()

    # Builder (for comparison)
    print("--- Builder Pattern (verbose) ---")
    svc3 = (ServiceBuilder(sender)
            .timeout(10)
            .retries(3)
            .verbose()
            .build())
    print("  builder:  ", end="")
    svc3.notify("charlie", "hi")
    print()

    print("Key: Python kwargs with defaults = Go functional options.")
    print("     Both solve 'constructor with many optional params'.")
    print("     Builder is redundant in Python but shown for comparison.")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/10_design_patterns_in_go/09_builder_vs_functional_options/main.go
package main

import (
	"fmt"
	"time"
)

// ============================================================
// APPROACH 1: Functional Options (Go-idiomatic)
// ============================================================

type Sender interface {
	Send(to, message string) error
}

type NotificationService struct {
	sender  Sender
	timeout time.Duration
	retries int
	verbose bool
}

// Option is a function that configures the service.
type Option func(*NotificationService)

func WithTimeout(d time.Duration) Option {
	return func(s *NotificationService) { s.timeout = d }
}

func WithRetries(n int) Option {
	return func(s *NotificationService) { s.retries = n }
}

func WithVerbose(v bool) Option {
	return func(s *NotificationService) { s.verbose = v }
}

// NewNotificationService uses functional options for optional config.
func NewNotificationService(sender Sender, opts ...Option) *NotificationService {
	// Defaults
	svc := &NotificationService{
		sender:  sender,
		timeout: 5 * time.Second,
		retries: 1,
		verbose: false,
	}
	// Apply options
	for _, opt := range opts {
		opt(svc)
	}
	return svc
}

func (s *NotificationService) Notify(to, msg string) {
	if s.verbose {
		fmt.Printf("  [VERBOSE] timeout=%s retries=%d\n", s.timeout, s.retries)
	}
	fmt.Printf("  Sending to %s: %s\n", to, msg)
}

// ============================================================
// APPROACH 2: Builder (less idiomatic in Go, shown for comparison)
// ============================================================

type ServiceBuilder struct {
	sender  Sender
	timeout time.Duration
	retries int
	verbose bool
}

func NewServiceBuilder(sender Sender) *ServiceBuilder {
	return &ServiceBuilder{
		sender:  sender,
		timeout: 5 * time.Second,
		retries: 1,
	}
}

func (b *ServiceBuilder) Timeout(d time.Duration) *ServiceBuilder {
	b.timeout = d
	return b
}

func (b *ServiceBuilder) Retries(n int) *ServiceBuilder {
	b.retries = n
	return b
}

func (b *ServiceBuilder) Verbose() *ServiceBuilder {
	b.verbose = true
	return b
}

func (b *ServiceBuilder) Build() *NotificationService {
	return &NotificationService{
		sender:  b.sender,
		timeout: b.timeout,
		retries: b.retries,
		verbose: b.verbose,
	}
}

// --- Simple sender for demo ---

type printSender struct{}

func (p *printSender) Send(to, message string) error {
	fmt.Printf("  [SEND] %s: %s\n", to, message)
	return nil
}

func main() {
	fmt.Println("=== Builder vs Functional Options ===")
	fmt.Println()
	sender := &printSender{}

	// Functional options (preferred)
	fmt.Println("--- Functional Options (Go-idiomatic) ---")

	svc1 := NewNotificationService(sender) // all defaults
	fmt.Print("  defaults: ")
	svc1.Notify("alice", "hello")

	svc2 := NewNotificationService(sender,
		WithTimeout(10*time.Second),
		WithRetries(3),
		WithVerbose(true),
	)
	fmt.Print("  custom:   ")
	svc2.Notify("bob", "world")
	fmt.Println()

	// Builder (for comparison)
	fmt.Println("--- Builder Pattern (less idiomatic) ---")
	svc3 := NewServiceBuilder(sender).
		Timeout(10 * time.Second).
		Retries(3).
		Verbose().
		Build()
	fmt.Print("  builder:  ")
	svc3.Notify("charlie", "hi")
	fmt.Println()

	fmt.Println("Key: functional options are Go-idiomatic for optional config.")
	fmt.Println("     Builder works but adds more ceremony with chained methods.")
}
